Document Nutanix (#1883)
spectrocloud/librarium: docs/docs-content/clusters/data-center/nutanix/register-nutanix-cloud.md

## Prerequisites

- A Palette account with system console access. The user with this privilege is the *admin user* of the self-hosted instance of [Palette](https://docs.spectrocloud.com/enterprise-version/system-management/#system-console) or [VerteX](https://docs.spectrocloud.com/vertex/system-management/#system-console).

- [`curl`](https://curl.se/docs/install.html) command installed.

<!-- - A valid Palette authentication token. To learn how to acquire an authentication token, review the [Authorization Token](https://docs.spectrocloud.com/user-management/authentication/authorization-token) guide. -->

## Setup

Use the following steps to prepare to register your cloud with Palette. 

### Customize YAML Configuration Files

1.  Download the following YAML files from the [Nutanix Cluster API (CAPI) Provider](https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix) GitHub repository:
    - **infrastructure-components.yaml**
    - **cluster-template.yaml**
  
  <br />

  :::caution

>   Review the [Nutanix](https://opendocs.nutanix.com/capx/v1.2.x/validated_integrations/#validated-versions) compatibility matrix to ensure you download the latest CAPI version of the files. 

Reviewer: We should have a requirement for the Nutanix CAPI version required, that must be compatible with our CAPI version.

Author: Added the second bullet, following the existing bullet about minimum supported CAPI version 1.5.3.

"- A Nutanix cloud created with the minimum supported CAPI version 1.5.3. Refer to the Nutanix [Validated Integrations](https://opendocs.nutanix.com/capx/v1.2.x/validated_integrations/#validated-versions) compatibility matrix.

- The Nutanix CAPI version must be compatibe with Palette's CAPI version."

  :::

  Issue the commands below to download the files.

    ```bash

>     curl -LO https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/releases/latest/download/cluster-template.yaml

Reviewer: Shouldn't be latest, no? Depends on the supported version

Author: Fixed as follows. We think this is clear.
![CleanShot 2023-12-20 at 14 25 11](https://github.com/spectrocloud/librarium/assets/117382432/fc2da682-75fc-499d-b4b0-3405d5c6d43c)

    

## Validate

Use the steps below to confirm you have the required files and verify the required sections are removed and modified. 

1. From your terminal, issue a command such as `ls -l` to list the files and confirm you have the following YAML templates:

    - infrastructure-components.yaml
    - cloudClusterTemplate.yaml    
    - controlPlanePoolTemplate.yaml
    - workerPoolTemplate.yaml

2. Ensure each template contains objects as listed in the table.

  | **Templates**                  | **Objects**           |
  |--------------------------------|------------------------|
  | **cloudClusterTemplate.yaml**    | ConfigMap<br />Secret<br />Cluster<br />NutanixCluster<br />MachineHealthCheck | 
  | **controlPlanePoolTemplate.yaml**| KubeadmControlPlane<br />NutanixMachineTemplate |
  | **workerPoolTemplate.yaml**     | KubeadmConfigTemplate<br />MachineDeployment<br />NutanixMachineTemplate |

3. Open each file and verify that all occurrences of `${NAMESPACE}` are removed. 

4. In the **controlPlanePoolTemplate.yaml** file, ensure `${CLUSTER_NAME}-cp-0` for the KubeadmControlPlane and NutanixMachineTemplate objects have the same name.

> 5. Verify parameters are modifed as described for each template in steps 6 and 7 of [Customize YAML Configuration Files](#customize-yaml-configuration-files).

Author: fixed typo

## Register the Cloud

Follow the steps below from your terminal to set the environment variables and invoke the APIs required to register a Nutanix cloud to Palette. Alternatively, you can use an API platform such as [Postman](https://www.postman.com/).

:::caution

> During the registration process, the template files and logo file must be located in the same directory where you execute the curl commands in the steps below. The logo file must not exceed 100KB in size.

Reviewer: And the recommend size in pixels? And ratio?

Reviewer 2: The engineering team said that there are no requirements as of now. Only size is limited to 100KB max.

Author: Added: The logo file must not exceed 100KB in size. To ensure image quality ensure at least one dimension in either width or height is 40 pixels. It is preferable that the image be transparent.

  ```

8. Register the control plane pool template. 

  ```bash
  curl --location --request PUT "${ENDPOINT}v1/clouds/cloudTypes/${CLOUD_TYPE}/content/templates/controlPlanePoolTemplate" \
        --header "Content-Type: multipart/form-data" \
        --header "Cookie: Authorization=${TOKEN}" \
        --form "fileName=@${controlPlanePoolTemplate.yaml}"
  ```

9. Register the worker pool template. 

  ```bash
    curl --location --request PUT "${ENDPOINT}v1/clouds/cloudTypes/${CLOUD_TYPE}/content/templates/workerPoolTemplate" \
          --header "Content-Type: multipart/form-data" \
          --header "Cookie: Authorization=${TOKEN}" \
          --form "fileName=@${workerPoolTemplate.yaml}"
    ```

## Validate

Use the steps below to confirm that the Nutanix cloud is successfully registered in Palette.

1. Log in to [Palette](https://console.spectrocloud.com/) as a tenant admin.

2. Navigate to the **left Main Menu** and select **Tenant Settings**.

3. Next, on the **Tenant Settings Menu**, select **Cloud Accounts**.

4. Verify that the added Nutanix account section is listed. You may need to scroll to view the account.

## Next Steps

> Now that your cloud is successfully registered with Palette, the Private Cloud Gateway (PCG) must be deployed. For guidance, review the [Install Private Cloud Gateway](install-pcg.md) guide.

Reviewer: ```suggestion
Now that your cloud is successfully registered with Palette, a Self-hosted Private Cloud Gateway (PCG) must be deployed. For guidance, review the [Install Private Cloud Gateway](install-pcg.md) guide.
```

---
sidebar_label: "Register Nutanix Cloud"
title: "Register Nutanix Cloud"
description: "Learn how to register a Nutanix cloud in Palette."
hide_table_of_contents: false
sidebar_position: 5
tags: ["data center", "nutanix"]
---

A System Administrator registers the Nutanix cloud in Palette by invoking system-level APIs. These APIs provide specific cloud information, the cloud logo, and the key-value pairs required to add the cloud to Palette. They also enable uploading the YAML templates used to create the cluster, control plane, and worker nodes. This section provides instructions on how to use the APIs to register a Nutanix cloud to Palette.

> 

Reviewer 2: Drafted

## Prerequisites

> 

Reviewer 2: Added `curl` as a prerequisite and changed the commands from `wget` to `curl`.

> - A Palette account with system-level access.

Reviewer 2: Drafted

<!-- - A valid Palette authentication token. To learn how to acquire an authentication token, review the [Authorization Token](https://docs.spectrocloud.com/user-management/authentication/authorization-token) guide. -->

## Setup

Use the following steps to prepare to register your cloud with Palette. 

### Customize YAML Configuration Files

1.  Download the following YAML files from the [Nutanix Cluster API (CAPI) Provider](https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix) GitHub repository:
    - `infrastructure-components.yaml`

>     - `cluster-template.yaml`

Author: fixed

  
  <br />

  :::caution

>   Check the [Nutanix](https://opendocs.nutanix.com/capx/v1.2.x/validated_integrations/#validated-versions) compatibility matrix to ensure you download the latest CAPI version of the files. 

Author: applied

  :::

  Issue the commands below to download the files.

```bash
  wget https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/releases/latest/download/cluster-template.yaml
  wget https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/releases/latest/download/infrastructure-components.yaml
```

2. Create two copies of `cluster-template.yaml` and rename them so you have the following files in addition to the `infrastructure-components.yaml`:
    - `cloudClusterTemplate.yaml`
    - `controlPlanePoolTemplate.yaml`

>     - `workerPoolTemplate.yaml`

Author: fixed

  Use the following commands to copy and rename the files.

```bash
  cp cluster-template.yaml cloudClusterTemplate.yaml 
  cp cluster-template.yaml controlPlanePoolTemplate.yaml
  mv cluster-template.yaml workerPoolTemplate.yaml
```

> 3. Open the `cloudClusterTemplate.yaml`, `controlPlanePoolTemplate.yaml`, and `workerPoolTemplate.yaml` files in the editor of your choice.

Author: fixed

4. Modify the YAML files to remove sections so that only those sections listed in the table below remain in each file.

  :::tip

  When editing the YAMLs, it is helpful to collapse the `spec` section to help you identify the sections to remove.  

  :::

  | **Templates**                   | **Objects**           |
  |--------------------------------|------------------------|
  | `cloudClusterTemplate.yaml`    | ConfigMap<br />Secret<br />Cluster<br />NutanixCluster<br />MachineHealthCheck | 
  | `controlPlanePoolTemplate.yaml`| KubeadmControlPlane<br />NutanixMachineTemplate |

>   | `workerPoolTemplate.yaml`      | KubeadmConfigTemplate<br />MachineDeployment<br />NutanixMachineTemplate |

Author: fixed

5. In all three templates, remove all occurrences of `${NAMESPACE}`, as Palette provides its own namespace.

6. In controlPlanePoolTemplate, edit the template as follows:
  
    - In the KubeadmControlPlane object, rename `machineTemplate.name: ${CLUSTER_NAME}-mt-0` as `${CLUSTER_NAME}-cp-0`.  

    - In the NutanixMachineTemplate object:
      - Rename `name: ${CLUSTER_NAME}-mt-0` as `${CLUSTER_NAME}-cp-0`.
      - Change `providerID` to `nutanix://${CLUSTER_NAME}-m1-cp-0`.

      <br /> 

      :::caution
      The `${CLUSTER_NAME}-cp-0` parameters for the KubeadmControlPlane and NutanixMachineTemplate objects must have the same name.
      :::

> 7. In workerPoolTemplate.yaml, change `providerID` to `providerID: nutanix://$CLUSTER_NAME}-m1-mt-0` within the `NutanixMachineTemplate` object. 

Author: fixed/applied

    

## Validate

Use the steps below to confirm you have the required files and verify the required sections are removed and modified. 

> 1. From your terminal, confirm you have the following YAML templates:

Author: fixed

  | `controlPlanePoolTemplate.yaml`| KubeadmControlPlane<br />NutanixMachineTemplate |
  | `workerPoolTemplate.yaml`      | KubeadmConfigTemplate<br />MachineDeployment<br />NutanixMachineTemplate |

3. Open each file and verify that all occurrences of `${NAMESPACE}` are removed. 

4. In controlPlanePoolTemplate.yaml, ensure `${CLUSTER_NAME}-cp-0` for the KubeadmControlPlane and NutanixMachineTemplate objects have the same name.

5. Verify parameters are modifed as described for each template in steps 6 and 7 of [Customize YAML Configuration Files](#customize-yaml-configuration-files).

## Register the Cloud

Follow the steps below from your terminal to set the environment variables and invoke the APIs required to register a Nutanix cloud to Palette. Alternatively, you can use an API platform such as [Postman](https://www.postman.com/).

:::caution
During the registration process, the template files and logo file must be located in the same directory where you execute the curl commands in the steps below. The logo file must not exceed 100KB in size.

:::

1. Export your Palette URL and the cloud type you want to register as environment variables.

> ```bash

Reviewer 2: Fixed and I'm also waiting for some answers :)

export ENDPOINT="https://palette.example.com"

> export CLOUD_TYPE="nutanix-custom"

Reviewer 3: rename cloudType as `nutanix` instead as CSI I specific to this

```

2. To acquire a system administrator authentication token, log in to the Palette API by using the `/v1/auth/syslogin` endpoint. Use the `curl` command below and replace the URL with the custom domain URL assigned to Palette. Ensure you replace the credentials with your system console credentials.

```bash
curl --insecure --location "$ENDPOINT/v1/auth/syslogin" \
 --header 'Content-Type: application/json' \
 --data "{
  "password": "**********",
  "username": "**********"
 }"
```

  The output contains your authorization token.

```bash hideClipBoard
{
  "Authorization": "**********",
  "IsPasswordReset": true
}
```

3. Copy the authorization token, assign it to a `TOKEN` shell variable, and export it. Replace the authorization value below with the value from the output.

```bash
export TOKEN="**********"
```

4. Register the Nutanix cloud type to Palette by using the `/v1/clouds/cloudTypes/register` endpoint.

```bash

> curl --location --request POST "${ENDPOINT}/v1/clouds/cloudTypes/register" \

Reviewer 2: I copied it from the engineering shell script. I'll test it and fix it.

--header "Content-Type: application/json" \
--header "Cookie: Authorization=${TOKEN}" \
--data "{
    "metadata": {
        "annotations": {},
        "labels": {},
        "name": "${CLOUD_TYPE}"
    },
    "spec": {
        "displayName": "Nutanix",
        "isControlPlaneManaged": false
    }
}"
```

> 5. Upload the Nutanix cloud logo by using the `curl` command displayed below. 

Reviewer 2: Definitely. Removed :)

```bash
curl --location --request PUT "${ENDPOINT}/v1/clouds/cloudTypes/${CLOUD_TYPE}/logo" \

> --header "Cookie: Authorization=${TOKEN}" \

Reviewer 2: Fixed.

---
sidebar_label: "Register Nutanix Cloud"
title: "Register Nutanix Cloud"
description: "Learn how to register a Nutanix cloud in Palette."
hide_table_of_contents: false
sidebar_position: 5
tags: ["data center", "nutanix"]
---

A system administrator registers the Nutanix cloud in Palette by invoking system-level APIs. These APIs provide specific cloud information, the cloud logo, and the key-value pairs required to add the cloud to Palette. They also enable uploading the YAML templates used to create the cluster, control plane, and worker nodes. This section provides instructions on how to download and modify the YAML templates, as well as how to use the APIs to register a Nutanix cloud to Palette.

## Prerequisites

- A Nutanix cloud created with the minimum supported CAPI version 1.5.3. Refer to the Nutanix [Validated Integrations](https://opendocs.nutanix.com/capx/v1.2.x/validated_integrations/#validated-versions) compatibility matrix.

> - The Nutanix CAPI version must be compatibe with Palette's CAPI version.

Author: Removed sentence - this no longer applies.

---
sidebar_label: "Register Nutanix Cloud"
title: "Register Nutanix Cloud"
description: "Learn how to register a Nutanix cloud in Palette."
hide_table_of_contents: false
sidebar_position: 5
tags: ["data center", "nutanix"]
---

A system administrator registers the Nutanix cloud in Palette by invoking system-level APIs. These APIs provide specific cloud information, the cloud logo, and the key-value pairs required to add the cloud to Palette. They also enable uploading the YAML templates used to create the cluster, control plane, and worker nodes. This section provides instructions on how to download and modify the YAML templates, as well as how to use the APIs to register a Nutanix cloud to Palette.

## Prerequisites

> - A Nutanix cloud created with the minimum supported CAPI version 1.5.3. Refer to the Nutanix [Validated Integrations](https://opendocs.nutanix.com/capx/v1.2.x/validated_integrations/#validated-versions) compatibility matrix.

Author: Applied.

> - The Nutanix CAPI version must be compatibe with Palette's CAPI version.

Author: Applied.

> - A Cluster API (CAPI) image created for the Nutanix cloud platform. For guidance, refer to [Building CAPI Images for Nutanix Cloud Platform](https://image-builder.sigs.k8s.io/capi/providers/nutanix). Images can be found in the Nutanix Prism dashboard under **Compute & Storage**.

Reviewer: ```suggestion
- A Cluster API (CAPI) image created for the Nutanix cloud platform. For guidance, refer to [Building CAPI Images for Nutanix Cloud Platform](https://image-builder.sigs.k8s.io/capi/providers/nutanix). Images can be found in the Nutanix Prism Central dashboard under **Compute & Storage**.
```

Author: Applied.

- A Palette account with system console access. The user with this privilege is the *admin user* of the self-hosted [Palette](https://docs.spectrocloud.com/enterprise-version/system-management/#system-console) or [VerteX](https://docs.spectrocloud.com/vertex/system-management/#system-console) instance.

- A Nutanix logo downloaded. Review logo requirements in [Register the Cloud](#register-the-cloud).

> - [`curl`](https://curl.se/docs/install.html) command installed.

Reviewer: Or any way to run API command to Palette/VerteX

Author: Added: ... or the method of your choice to issue API commands for standard Palette and Palette VerteX.

<!-- - A valid Palette authentication token. To learn how to acquire an authentication token, review the [Authorization Token](https://docs.spectrocloud.com/user-management/authentication/authorization-token) guide. -->

## Setup

Use the following steps to prepare to register your cloud with Palette. 

### Customize YAML Configuration Files

> 1.  Download the following YAML files from the [Nutanix Cluster API (CAPI) Provider](https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix) GitHub repository:

Reviewer: Those should actually be downloaded from a specific version of CAPX, in the [release page](https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/releases). For example for 1.2.4, you scroll down to the [assets](https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/releases/tag/v1.2.4). This is important!

<img width="1234" alt="Screenshot 2023-12-21 at 18 15 33" src="https://github.com/spectrocloud/librarium/assets/10382023/4a7eb5e8-936e-4bec-ad71-763e1a82d17b">

Author: Removed curl commands and revised step 1 to say:

1.  Download the following YAML files from a specific version of CAPX found on the [Nutanix Releases](https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/releases) page in the GitHub repository.

Reviewer 2: I think it is a bit confusing now. Which specific version of CAPX? I liked the curl commands better.
What if we create an env variable to store the CAPx version and keep the curl commands?

:::

1. Export the URL of your self-hosted Palette or VerteX instance and the cloud type as environment variables. Additionally, export the path to the YAML templates and logo file.

  ```bash
  export ENDPOINT="https://palette.example.com"
  export CLOUD_TYPE="nutanix"
  export cloudLogo="/path/to/the/file/cloud-logo.png"
  export infraComponents="/path/to/the/file/infrastructure-components.yaml"
  export cloudClusterTemplate="/path/to/the/file/cloudClusterTemplate.yaml"
  export  controlPlanePoolTemplate="/path/to/the/file/controlPlanePoolTemplate.yaml"
  export workerPoolTemplate="/path/to/the/file/workerPoolTemplate.yaml"
  ```

  :::caution
  The CLOUD_TYPE variable value must be set as `nutanix`, as this value will be used for the `name` value in the `/v1/clouds/cloudTypes/register` cloud registration API in step 4 below. Setting `name` as `nutanix` will make the out-of-the-box **Nutanix CSI** pack available to users when they create a cluster profile in Palette. 
  :::

2. To acquire system administrator credentials, use the `/v1/auth/syslogin` endpoint. Issue the `curl` command below and ensure you replace the credentials with your system console credentials.

  ```bash

>   curl --insecure --location "$ENDPOINT/v1/auth/syslogin" \

Reviewer: We should **not** provide examples where we have `-k` or `--insecure` in our docs, righ @karl-cardenas-coding ?

Author: Removed `--insecure`.

Also changed double quotes to single quotes in the --data block, as discussed.

  --header 'Content-Type: application/json' \
  --header 'Accept: application/json' \
  --data "{
    "password": "**********",
    "username": "**********"
  }"
  ```

  The output contains your authorization token.

  ```bash hideClipBoard
  {
    "Authorization": "**********",
    "IsPasswordReset": true
  }
  ```

3. Copy the authorization token, assign it to a `TOKEN` shell variable, and export it. Replace the authorization value below with the value from the output.

  ```bash
  export TOKEN="**********"
  ```

4. Register the Nutanix cloud type in Palette using the `/v1/clouds/cloudTypes/register` endpoint.

  ```bash
  curl --location --request POST "${ENDPOINT}/v1/clouds/cloudTypes/register" \
  --header "Content-Type: application/json" \
  --header "Authorization: ${TOKEN}" \

>   --data "{

Reviewer: ```suggestion
  --data '{
```

Author: Applied.

      "metadata": {
          "annotations": {},
          "labels": {},
          "name": "${CLOUD_TYPE}"
      },
      "spec": {
          "displayName": "Nutanix",
          "isControlPlaneManaged": false
      }

>   }"

Reviewer: ```suggestion
  }'
```

Author: Applied

---
sidebar_label: "Register Nutanix Cloud"
title: "Register Nutanix Cloud"
description: "Learn how to register a Nutanix cloud in Palette."
hide_table_of_contents: false
sidebar_position: 5
tags: ["data center", "nutanix"]
---

A System Administrator registers the Nutanix cloud in Palette by invoking system-level APIs. These APIs provide specific cloud information, the cloud logo, and the key-value pairs required to add the cloud to Palette. They also enable uploading the YAML templates used to create the cluster, control plane, and worker nodes. This section provides instructions on how to use the APIs to register a Nutanix cloud to Palette.

## Prerequisites

- A Nutanix Prism Central account.

- A Palette account with system-level access.

- A valid Palette authentication token. To learn how to acquire an authentication token, review the [Authorization Token](https://docs.spectrocloud.com/user-management/authentication/authorization-token) guide.

> - Downloaded infrastructure-components.yaml from the Nutanix space in GitHub. The YAML contains all the Custom Resource Definitions (CRDs) for Nutanix cluster resources.

Author: fixed

> - YAML templates, which can be found on the cloud provider release pages in GitHub:

Author: removed

>   - control-plane template

Author: removed

  - worker template

>   - cluster template (cluster-template.yaml)

Author: removed

  The `${CLUSTER_NAME}-cp-0` parameters for the KubeadmControlPlane and NutanixMachineTemplate objects must have the same name.
  :::

8. In **workerPoolTemplate.yaml**, change `providerID` to `providerID: nutanix://${CLUSTER_NAME}-m1-mt-0` within the `NutanixMachineTemplate` object.

  :::caution

>   The following modifications are only applicable to VerteX instances.  

Reviewer 4: ```suggestion
  The following modifications in steps 9 and 10 only apply to VerteX instances.  
```
Just to make it very clear it's the two steps below.
